htmresearch/algorithms/column_pooler.py

```diff
@@ -174,16 +192,127 @@ def compute(self, feedforwardInput=(), lateralInputs=(),
 
     @param  learn (bool)
             If True, we are learning a new object
-    """
 
+    @param predictedInput (sequence)
+           Sorted indices of active feedforward input bits
+    """
     if feedforwardGrowthCandidates is None:
       feedforwardGrowthCandidates = feedforwardInput
 
-    if learn:
+    if not learn:
+      self._computeInferenceMode(feedforwardInput, lateralInputs)
+
+    elif not self.onlineLearning:
       self._computeLearningMode(feedforwardInput, lateralInputs,
                                 feedforwardGrowthCandidates)
     else:
-      self._computeInferenceMode(feedforwardInput, lateralInputs)
+      if (predictedInput is not None and
+          len(predictedInput) > self.predictedInhibitionThreshold):
+        predictedActiveInput = numpy.intersect1d(feedforwardInput, predictedInput)
+        predictedGrowthCandidates = numpy.intersect1d(feedforwardGrowthCandidates, predictedInput)
+        self._computeInferenceMode(predictedActiveInput, lateralInputs)
+        self._computeOnlineLearningMode(predictedActiveInput, lateralInputs,
+                                        feedforwardGrowthCandidates)
+      elif (numpy.abs(len(self.activeCells) - self.sdrSize)
+          > (self.learningTolerance*self.sdrSize)):
+        # If the pooler doesn't have a single representation, try to infer one,
+        # before actually attempting to learn.
+        self._computeInferenceMode(feedforwardInput, lateralInputs)
+        self._computeOnlineLearningMode(feedforwardInput, lateralInputs,
+                                        feedforwardGrowthCandidates)
+      else:
+        # If there isn't predicted input and we have a single SDR,
+        # we are extending that representation and should just learn.
+        self._computeOnlineLearningMode(feedforwardInput, lateralInputs,
+                                  feedforwardGrowthCandidates)
+        self._computeInferenceMode(feedforwardInput, lateralInputs)
```

Comment:
Why do we infer here? `_computeOnlineLearningMode` should have already selected the active cells.